src/Service/ServiceExe.cs

```diff
@@ -1791,42 +1791,58 @@ public async Task<bool> TryVerifyContainerConnection()
+            w.Start();
 
-            try
+            while(true)
             {
-                if (ContainerGuid == Guid.Empty)
+                try
                 {
-                    // If no container info has been found yet, double-check check the registry to
-                    // ensure we never fail a container Task or command due to a race condition.
-                    UpdateContainerInfo();
-                }
+                    if (ContainerGuid == Guid.Empty)
+                    {
+                        // If no container info has been found yet, double-check check the registry to
+                        // ensure we never fail a container Task or command due to a race condition.
+                        UpdateContainerInfo();
+                    }
 
-                await ConnectToContainer();
+                    await ConnectToContainer();
 
-                if (previousContainerStatus == IsContainerConnected)
-                {
-                    // Verify it is still working properly
-                    await _containerClient.GetServiceVersionString();
-                }
-                else
-                {
-                    // Initial connection made or reconnected
-                    LogServiceEvent(new ServiceEvent(ServiceEventType.ContainerConnected, ContainerGuid, Resources.ContainerConnected));
+                    if (previousContainerStatus == IsContainerConnected)
+                    {
+                        // Verify it is still working properly
+                        await _containerClient.GetServiceVersionString();
+                    }
+                    else
+                    {
+                        // Initial connection made or reconnected
+                        LogServiceEvent(new ServiceEvent(ServiceEventType.ContainerConnected, ContainerGuid, Resources.ContainerConnected));
+                    }
+
+                    // Connection established, exit loop
+                    break;
                 }
-            }
-            catch (FactoryOrchestratorContainerException)
-            {
-                if (previousContainerStatus == IsContainerConnected)
+                catch (FactoryOrchestratorContainerException)
                 {
-                    // Connection lost
-                    LogServiceEvent(new ServiceEvent(ServiceEventType.ContainerDisconnected, previousContainerGuid, Resources.ContainerDisconnected));
-                }
+                    if (previousContainerStatus == IsContainerConnected)
+                    {
+                        // Connection lost
+                        LogServiceEvent(new ServiceEvent(ServiceEventType.ContainerDisconnected, previousContainerGuid, Resources.ContainerDisconnected));
+                    }
 
-                throw;
+                    if (w.ElapsedMilliseconds * 1000 > retrySeconds)
+                    {
+                        throw;

```

Comment:
would we want to throw a timeout exception here?

Reply:
No, I only want container APIs to throw exceptions of type FactoryOrchestratorContainerException. I think the fact that it timed out is more of an implementation detail that a consumer doesn't need to know.

Reply:
Ah gotcha, makes sense